overlord/devicestate/devicemgr.go

```diff
@@ -288,8 +291,20 @@ func (m *DeviceManager) SystemMode(sysExpect SysExpectation) string {
 
 // StartUp implements StateStarterUp.Startup.
 func (m *DeviceManager) StartUp() error {
-	if m.shouldMountUbuntuSave() {
-		if err := m.setupUbuntuSave(); err != nil {
+	m.state.Lock()
+	defer m.state.Unlock()
+	defer m.earlyCleanup()
+
+	var dev snapstate.DeviceContext
+	dev, err := m.earlyDeviceContext()
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+	// if ErrNoState and dev is nil assume classic system here,
+	// any error will re-surface again in the main first boot code
+
+	if m.shouldMountUbuntuSave(dev) {
```

Comment:
nitpick: I'd rather have
```suggestion
	if dev != nil && m.shouldMountUbuntuSave(dev) {
```
here and assume that `shouldMountUbuntuSave()` needs a valid device; to me it feels more logical and makes the code here more readable (the reader would not need to look at `shouldMountUbuntuSave()` implementation to convince himself that the function properly handles the case where `dev` is null).

Reply:
changed